### PR TITLE
write rate limits to redis

### PR DESCRIPTION
**Pull request by alrocar**
#13088

### CODE REVIEW

File: db/migrate/20180227084750_create_account_types.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    create_table :account_types do
+      String      :account_type, null: false
+      foreign_key :rate_limit_id, :rate_limits, type: :uuid, on_delete: :cascade, null: false
```
Comment (javitonino) [inline db/migrate/20180227084750_create_account_types.rb]: The `CASCADE` is very dangerous here. If you delete a rate limit, it will delete the plan, so a lot of users may be left without the default limits. I'd rather have it report an error than silently delete cascading. Maybe too cautious?

File: spec/factories/account_type.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :account_type_free, class: Carto::AccountType do
+    account_type 'FREE'
+  end
```
Comment (javitonino) [inline spec/factories/account_type.rb]: You could add an association to this factory to default to creating some rate_limits: https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -2672,6 +2672,100 @@ def enabled_api_key?(api_key)
...
+      $limits_metadata.EXISTS("#{@map_prefix}named_get").should == 0
+      $limits_metadata.EXISTS("#{@map_prefix}named").should == 0
+      $limits_metadata.EXISTS("#{@map_prefix}named_update").should == 0
+      $limits_metadata.EXISTS("#{@map_prefix}named_delete").should == 0
+      $limits_metadata.EXISTS("#{@map_prefix}named_tiles").should == 0
+    end
+
+    it 'creates rate limits from user account type' do
+      FactoryGirl.create(:account_type_free, rate_limit_id: @rate_limits.id)
+      @user = FactoryGirl.create(:valid_user)
```
Comment (javitonino) [inline spec/models/user_spec.rb]: A user is already created in `before(:each)`. This is overriding that user which is double work for nothing, and also, the original user won't be deleted correctly.

File: app/services/carto/user_metadata_export_service.rb [hunk 4]
```diff
@@ -26,7 +26,7 @@ module UserMetadataExportServiceConfiguration
       :salesforce_datasource_enabled, :builder_enabled, :geocoder_provider, :isolines_provider, :routing_provider,
       :github_user_id, :engine_enabled, :mapzen_routing_quota, :mapzen_routing_block_price, :soft_mapzen_routing_limit,
       :no_map_logo, :org_admin, :last_name, :user_render_timeout, :database_render_timeout, :frontend_version,
-      :asset_host, :state
+      :asset_host, :state, :rate_limit_id
```
Comment (javitonino) [inline app/services/carto/user_metadata_export_service.rb]: This just exports the ID, which is worthless when restoring. We have to export the actual limits instead. We can leave this for another PR though.

So:
1. Create the ticket for adding the rate limits to the user migrations.
2. Since this attribute will never get exported directly, you can ignore it at the test that checks all attributes are exported. Remove it from here and tweak the test to pass.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -2672,6 +2672,100 @@ def enabled_api_key?(api_key)
...
+      @map_prefix = "limits:rate:store:#{@user.username}:maps:"
+    end
+
+    after :each do
+      @user.destroy unless @user.nil?
+      @rate_limits.destroy unless @rate_limits.nil?
+      @rate_limits2.destroy unless @rate_limits2.nil?
+    end
+
+    it 'does not create rate limits when account_type does not exist' do
```
Comment (javitonino) [inline spec/models/user_spec.rb]: This should never happen. I'll write about the deployment plan at the issue.

**Comment by javitonino** [comment]
Let's talk about how to deploy this. I would like to create a FK from `users` to `account_type`. This probably means that we have to split the deploy in many steps:

1. Create new tables
2. Populate account_type table (and corresponding rate_limits entries)
3. Add the users->account_type FK
4. Deploy the code

Did I miss something?

File: app/models/carto/rate_limit.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    serialize :maps_analysis, RateLimitValues
+    serialize :maps_tile, RateLimitValues
+    serialize :maps_attributes, RateLimitValues
+    serialize :maps_named_list, RateLimitValues
+    serialize :maps_named_create, RateLimitValues
+    serialize :maps_named_get, RateLimitValues
+    serialize :maps_named, RateLimitValues
+    serialize :maps_named_update, RateLimitValues
+    serialize :maps_named_delete, RateLimitValues
+    serialize :maps_named_tiles, RateLimitValues
```
[the remark this answers is not in the document]
Comment (alrocar) [reply app/models/carto/rate_limit.rb]: good point, thx :)

File: db/migrate/20180227084750_create_account_types.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    create_table :account_types do
```
[the remark this answers is not in the document]
Comment (alrocar) [reply db/migrate/20180227084750_create_account_types.rb]: I'd use account_type as the PK

File: db/migrate/20180227084750_create_account_types.rb (hunk 1 above)
Comment (alrocar) [reply db/migrate/20180227084750_create_account_types.rb]: seems good, adding `on_delete: :restrict`

File: spec/models/user_spec.rb (hunk 5 above)
Comment (alrocar) [reply spec/models/user_spec.rb]: removing the test

File: app/models/user.rb [hunk 8]
```diff
@@ -994,6 +994,26 @@ def save_metadata
...
+    rate_limits.to_redis.each do |key, value|
+      $limits_metadata.DEL "limits:rate:store:#{username}:#{key}"
+      $limits_metadata.LPUSH "limits:rate:store:#{username}:#{key}", value
+    end
+  rescue ActiveRecord::RecordNotFound => e
+    CartoDB.notify_exception(e, user: self)
+  end
+
+  def get_rate_limits
+    rate_limits_id = rate_limit_id ||
```
Comment (javitonino) [inline app/models/user.rb]: Some ideas in my sleep. This should probably be the owner's `account_type` for organizations. The reason being that all non-owner users will have this set to `ORGANIZATION USER`.

But I think we want to be able to override per-user. So I'd say, fallback in this order:
- `user.rate_limit`
- `user.organization.owner.account_type.rate_limit if organization_user?`
- `user.account_type.rate_limit`

File: app/models/carto/rate_limit_values.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    def to_redis_array
+      RateLimitValues.dump(self)
+    end
+
+    def self.dump(rate_limit_values)
+      return [] if rate_limit_values.nil?
+
+      result = []
+      rate_limit_values.rate_limits.reverse.each do |rate_limit|
```
[the remark this answers is not in the document]
Comment (alrocar) [reply app/models/carto/rate_limit_values.rb]: mmh I was using `LPUSH` which inserts at the head... missed I just can use RPUSH and avoid reversing the list (also the rate_limits where stored reversed in the database...)

Thanks for pointing this out!

File: app/models/user.rb (hunk 8 above)
Comment (alrocar) [reply app/models/user.rb]: what if the org owner has a custom rate_limit, should we consider it instead of using the account_type?

Comment (javitonino) [reply app/models/user.rb]: IMO, no. Since the `rate_limits` are at the user model, we should only use the one for the current user. If, in the future, we want to override the limits for the entire org, we should add a `rate_limit_id` to the organization model.

**Comment by alrocar** [comment]
The deployment plan seems good to me 👍

File: app/models/user.rb [hunk 10]
```diff
@@ -994,6 +994,26 @@ def save_metadata
                           'db_public',                 database_timeout,
                           'render',                    user_render_timeout,
                           'render_public',             database_render_timeout
+    save_rate_limits
+  end
+
+  def save_rate_limits
+    rate_limits.save_to_redis(self)
+  rescue ActiveRecord::RecordNotFound => e
+    CartoDB.notify_exception(e, user: self)
```
Comment (javitonino) [inline app/models/user.rb]: Use the new logger interface:

`CartoDB::Logger.error(message: 'Error retrieving user rate limits', exception: e)`

This should probably be in the `rate_limits` method though, to avoid putting a `rescue` in all of it's callers.

File: app/models/user.rb [hunk 11]
```diff
@@ -994,6 +994,26 @@ def save_metadata
...
+  end
+
+  def rate_limits
```
Comment (javitonino) [inline app/models/user.rb]: I feel this methods does too much. Let's try to split it into "associations" and "logic". If we used a single ORM this would be easier to see, but let's try to replicate the same anyway (as we will eventually migrate the user model to AR).

- One method to return the `AccountType` model of the user. No more logic (just `Carto::AccountType.find(account_type)`. This behaves like an association
- Another method to return the effective account type (i.e: handles organization fallback to the owner). This is logic. You may even consider putting a shortcut for this at the organization model.
- A method to return the `rate_limit` of the user. Just a `Carto::RateLimit.find(rate_limit_id)`. It behaves like an association.
- A method to return the effective rate limit. This is basically `rate_limit || effective_account_type.rate_limit`. This is the logic part.

That way, you have four, one-liner, methods, which should make this easier to read.

File: app/models/carto/rate_limit_values.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    def to_redis_array
+      RateLimitValues.dump(self)
+    end
+
+    def self.dump(rate_limit_values)
+      return [] if rate_limit_values.nil?
+
+      result = []
+      rate_limit_values.each do |rate_limit|
```
Comment (javitonino) [inline app/models/carto/rate_limit_values.rb]: Consider [`flat_map`](https://apidock.com/ruby/Enumerable/flat_map). 

You can also move the order in which the values are dumped to `RateLimitValue`. It makes sense because that class does the match between array positions and variables in `initialize`. This way, the "knowledge" of which order the parameters are resides in a single class `RateLimitValue`.

All together:
```ruby
rate_limit_values.flat_map(&:to_array)
```

```ruby
class RateLimitValue
  def to_array
    [max_burst, count_per_period, period]
  end
end
```

File: app/models/carto/rate_limit_values.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    extend Forwardable
+    delegate [].methods => :@rate_limits
+
+    VALUES_PER_RATE_LIMIT = 3
+
+    def initialize(values)
+      @rate_limits = []
+      values = convert_to_array(values)
+
+      return if !values || values.length % VALUES_PER_RATE_LIMIT != 0
```
Comment (javitonino) [inline app/models/carto/rate_limit_values.rb]: Be aggressive, raise an error! :boom:

File: app/models/carto/rate_limit_values.rb [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    end
+  end
+
+  class RateLimitValue
+    attr_accessor :max_burst, :count_per_period, :period
+
+    def initialize(values)
+      self.max_burst = values[0].to_i
+      self.count_per_period = values[1].to_i
+      self.period = values[2].to_i
```
Comment (javitonino) [inline app/models/carto/rate_limit_values.rb]: This is just so you can be a cool guy xD. You can use the [multiple assignment syntax](https://docs.ruby-lang.org/en/2.2.0/syntax/assignment_rdoc.html#label-Multiple+Assignment) here:
`self.max_burst, self.count_per_period, self.period = values.map(&:to_i)`

File: app/models/user.rb (hunk 11 above)
Comment (alrocar) [reply app/models/user.rb]: `neat`

**Comment by alrocar** [comment]
I had to do some changes while writing the tests:

- RateLimitValues does not delegate all the `[]` methods, that was giving problems such as `RateLimitValues` was returned as an `Array` and there was no access to the class methods.
- I'm calling at `as_rate_limit_values` to force validation of the rate limits.
- After persisting the a RateLimit instance I'm reserializing to RateLimitValues

**Comment by javitonino** [comment]
Acceptance issues:
- If a limit is empty (`[]`) the save to Redis fails: `ERR wrong number of arguments for 'rpush' command`. We should validate that all limits have data. In fact, we can validate that the number of elements in each array is >=3 and multiple of 3.
- Could we delegate the `:[]`  accessor to the array, so we can do `rate_limit.maps_anonymous[0]`? Also `:<<`, `:[]=` and `clear`. I tried to use all of these during acceptance, so they may be useful for more people.

**Comment by javitonino** [comment]
Also, we got asked to create a FF to avoid setting limits if the user doesn't have the feature enabled. Avoiding the save to redis should be enough.

File: app/models/carto/rate_limit_values.rb [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+      Carto::InsertableArray.new(rate_limit_values.flat_map(&:to_array))
+    end
+
+    def self.load(values)
+      RateLimitValues.new(values)
+    end
+
+    def validate
+      raise 'Error. Empty rate limits for endpoint' if @rate_limits.empty?
```
Comment (javitonino) [inline app/models/carto/rate_limit_values.rb]: Let's use ActiveRecord validations: http://edgeguides.rubyonrails.org/active_record_validations.html

The idea would be to move the validation to the model. Since this is an array, you could delegate `present?` and use the presence validator: http://edgeguides.rubyonrails.org/active_record_validations.html#presence to ensure it's not empty.

**Comment by javitonino** [comment]
Very nice, let's start the deploy party!
